### ISSUE TITLE
nameserver changed

### ISSUE THREAD

**Issue by bkmagnetron** [openshift/openshift-ansible#2163]
In my private network I have deployed a bind server with IP 192.168.0.88. 

**Before Installing:**

```
# node1.localdomain 192.168.0.101
$ sudo cat /etc/resolv.conf 
search localdomain
nameserver 192.168.0.88
nameserver 8.8.8.8
```

```
# docker options
OPTIONS=' --selinux-enabled --dns 192.168.0.88 --dns 8.8.8.8 --insecure-registry=172.30.0.0/16 --log-driver=json-file'
```

**After Installing:**
_nameserver_ has been changed.

```
# node1.localdomain 192.168.0.101
$ sudo cat /etc/resolv.conf 
search localdomain
nameserver 192.168.0.101
nameserver 8.8.8.8
```

_dns_ options were removed

```
# docker options
OPTIONS=' --selinux-enabled --insecure-registry=172.30.0.0/16 --log-driver=json-file --log-opt max-size=50m'
```

**Installation:**

```
# cat /etc/ansible/hosts 
# Create an OSEv3 group that contains the masters and nodes groups
[OSEv3:children]
masters
nodes

# Set variables common for all OSEv3 hosts
[OSEv3:vars]
# SSH user, this user should allow ssh based auth without requiring a password
ansible_ssh_user=root

# If ansible_ssh_user is not root, ansible_sudo must be set to true
#ansible_sudo=true

deployment_type=origin

# uncomment the following to enable htpasswd authentication; defaults to DenyAllPasswordIdentityProvider
#openshift_master_identity_providers=[{'name': 'htpasswd_auth', 'login': 'true', 'challenge': 'true', 'kind': 'HTPasswdPasswordIdentityProvider', 'filename': '/etc/origin/master/htpasswd'}]

# host group for masters
[masters]
node1.localdomain

# host group for nodes, includes region info
[nodes]
node1.localdomain openshift_node_labels="{'region': 'infra', 'zone': 'default'}" openshift_schedulable=true
```

```
# ansible-playbook ~/openshift-ansible/playbooks/byo/config.yml
```

**Issue:**

```
I0717 06:30:03.649386       1 builder.go:57] Master version "v1.2.0", Builder version "v1.2.0"
I0717 06:30:03.716060       1 builder.go:145] Running build with cgroup limits: api.CGroupLimits{MemoryLimitBytes:92233720368547, CPUShares:2, CPUPeriod:100000, CPUQuota:-1, MemorySwap:92233720368547}
I0717 06:30:03.890806       1 sti.go:205] The value of ALLOWED_UIDS is [1-]
I0717 06:30:03.890843       1 sti.go:213] The value of DROP_CAPS is [KILL,MKNOD,SETGID,SETUID,SYS_CHROOT]
I0717 06:30:03.902074       1 docker.go:355] Using locally available image "centos/python-34-centos7@sha256:e3ad517f317ad7a5d1939ab38e406825d2b51e9cc676df53825356fc18027c15"
I0717 06:30:03.906015       1 sti.go:232] Creating a new S2I builder with build config: "Builder Name:\t\t\tPython 3.4\nBuilder Image:\t\t\tcentos/python-34-centos7@sha256:e3ad517f317ad7a5d1939ab38e406825d2b51e9cc676df53825356fc18027c15\nBuilder Image Version:\t\te761571bb1bc910bda65674e0d43c4c3509a7823\nBuilder Base Version:\t\te178f1c\nSource:\t\t\t\tfile:///tmp/s2i-build680461441/upload/src\nOutput Image Tag:\t\tdjango-ex/django-example-2:ffd6400a\nEnvironment:\t\t\tOPENSHIFT_BUILD_NAME=django-example-2,OPENSHIFT_BUILD_NAMESPACE=django-ex,OPENSHIFT_BUILD_SOURCE=https://github.com/openshift/django-ex.git,PIP_INDEX_URL=\nIncremental Build:\t\tdisabled\nRemove Old Build:\t\tdisabled\nBuilder Pull Policy:\t\tif-not-present\nPrevious Image Pull Policy:\talways\nQuiet:\t\t\t\tdisabled\nLayered Build:\t\t\tdisabled\nWorkdir:\t\t\t/tmp/s2i-build680461441\nDocker NetworkMode:\t\tcontainer:3f83a7f33dd9a756ef3abc9939c6a51e22d0519bc97937403d75a9591d167843\nDocker Endpoint:\t\tunix:///var/run/docker.sock\n"
I0717 06:30:03.912004       1 docker.go:355] Using locally available image "centos/python-34-centos7@sha256:e3ad517f317ad7a5d1939ab38e406825d2b51e9cc676df53825356fc18027c15"
I0717 06:30:03.923626       1 docker.go:355] Using locally available image "centos/python-34-centos7@sha256:e3ad517f317ad7a5d1939ab38e406825d2b51e9cc676df53825356fc18027c15"
I0717 06:30:03.923688       1 docker.go:475] Image contains io.openshift.s2i.scripts-url set to 'image:///usr/libexec/s2i'
I0717 06:30:03.923751       1 sti.go:140] Preparing to build django-ex/django-example-2:ffd6400a
I0717 06:30:04.081751       1 source.go:197] Downloading "https://github.com/openshift/django-ex.git" ...
I0717 06:30:20.081999       1 cleanup.go:23] Removing temporary directory /tmp/s2i-build680461441
I0717 06:30:20.082028       1 fs.go:156] Removing directory '/tmp/s2i-build680461441'
F0717 06:30:20.082433       1 builder.go:204] Error: build error: timeout while waiting for remote repository "https://github.com/openshift/django-ex.git"
```

**How I solved it:**
I again appended `--dns 192.168.0.88 --dns 8.8.8.8` to docker `OPTIONS` again. And restarted the node(tried restarting the docker alone but it didn't worked).

**Comment by sdodson** [comment]
The installer will configure dnsmasq to forward '*.cluster.local' to skydns and all other dns requests to the otherwise default nameservers. It will then reconfigure /etc/resolv.conf to use dnsmasq. We do have a known issue where if the default interface on the host is not NM_CONTROLLED=yes in /etc/sysconfig/network-scripts/ifcfg-* then the setup will fail. Is that the case on your installation? IF so can you set it to yes?

**Comment by bkmagnetron** [comment]
@sdodson Thanks `NM_CONTROLLED="yes"` fixed the problem and the django app that I deployed also worked fine i.e. route created and can able to do `curl django-ex.my-domain` . But when I **restart** both the master and nodes. I get the following in events
```
Readiness probe failed: Get http://10.1.0.2:8080/: read tcp 10.1.0.2:8080: use of closed network connection
```
```
Liveness probe failed: Get http://10.1.0.2:8080/: read tcp 10.1.0.2:8080: use of closed network connection
```
10.1.0.2 is the ip of the pod.
I can able to `curl 10.1.0.2:8080` but not `curl django-ex.my-domain`

I don't know weather this relates to the environment setup by ansible or the issue with kubernetes.

**Comment by sdodson** [comment]
Django-ex.my-domain. is the route's hostname? If so that would depend on
your wildcard DNS setup and is unrelated to dnsmasq.

**Comment by bkmagnetron** [comment]
@sdodson I did used wild card DNS. The router URL to Django is not actually `django-ex.my-domain`. It is actually `django-ex.10.28.27.66.xip.io` which actually works like wildcard DNS. And `10.28.27.66` is the system where router and docker registry were deployed which is actually master(openshift_schedulable=true).

**Comment by bkmagnetron** [comment]
**Briefing:**
After installing OpenShift using Ansible I create a snapshot of the two machines (master and node in VirtualBox) in **running** state.  After successfully deploying django-ex I created a another snapshot in **shutdown** state.

Then, I booted up both the machines(state changed) from the last snapshot that leads to this issues thread.

Now, I used **Restore Snapshot** of the **shutdown** state(way to issue in this thread) of both machines and booted up. _But_, this time it booted up without `Readiness probe failed` and `Liveness probe failed`. And after waiting few minutes(approx 10-15min) `curl django-ex.10.28.27.66.xip.io` **worked**.

**Conclusion:**
So, what I think is that there is might be some problem is my private DNS because that's the only thing which is not changed when I used **Restore Snapshot**. So, if I'm right this issue thread can be closed.

Note: Before using **Restore Snapshot** i.e. state where this issue happened I tried scaling down the router pod to **0** and that got struck at Terminating pod. I scaled up again a new pod is created while the other one is still **Terminating** but the new pod also had the same issue. Then only I decided to use **Restore Snapshot**.

**Comment by sdodson** [comment]
Ok, I suspect there was just some delay in the router coming online. Do you think we can close this?

**Comment by bkmagnetron** [comment]
Manually adding `NM_CONTROLLED="yes"` solved the major issue. But it will be good to add a FAIL check in Ansible or configure it automatically.

**Comment by detiber** [comment]
I don't think we want to enable it by default, but I do agree with failing if it isn't network manager controlled.

**Comment by EddFigueiredo** [comment]
I'm having the exact same problem... Already add NM_CONTROLLED=yes to my interface... Add the --dns to docker file... But I'm unable to create apps.

`I0805 16:07:57.929856       1 builder.go:57] Master version "v1.2.1", Builder version "v1.2.1"
I0805 16:07:57.934301       1 builder.go:145] Running build with cgroup limits: api.CGroupLimits{MemoryLimitBytes:92233720368547, CPUShares:2, CPUPeriod:100000, CPUQuota:-1, MemorySwap:92233720368547}
I0805 16:07:57.940164       1 sti.go:205] The value of ALLOWED_UIDS is [1-]
I0805 16:07:57.940197       1 sti.go:213] The value of DROP_CAPS is [KILL,MKNOD,SETGID,SETUID,SYS_CHROOT]
I0805 16:07:57.944790       1 docker.go:351] Image "openshift/nodejs-010-centos7@sha256:851ec2ace8a73a9cc3868edcbff1ac50d5f382d5079b7bfa4e7ff2bb531c171f" not available locally, pulling ...
I0805 16:07:57.944855       1 docker.go:373] Pulling Docker image openshift/nodejs-010-centos7@sha256:851ec2ace8a73a9cc3868edcbff1ac50d5f382d5079b7bfa4e7ff2bb531c171f ...
I0805 16:09:38.867149       1 sti.go:232] Creating a new S2I builder with build config: "Builder Name:\t\t\tNode.js 0.10\nBuilder Image:\t\t\topenshift/nodejs-010-centos7@sha256:851ec2ace8a73a9cc3868edcbff1ac50d5f382d5079b7bfa4e7ff2bb531c171f\nBuilder Image Version:\t\t35c861b683db01219b9a00f6400505f263e0ad18\nBuilder Base Version:\t\te178f1c\nSource:\t\t\t\tfile:///tmp/s2i-build760427484/upload/src#master\nOutput Image Tag:\t\tzupme/nodejs-teste-1:dcfd037c\nEnvironment:\t\t\tOPENSHIFT_BUILD_NAME=nodejs-teste-1,OPENSHIFT_BUILD_NAMESPACE=zupme,OPENSHIFT_BUILD_SOURCE=https://github.com/openshift/nodejs-ex.git,OPENSHIFT_BUILD_REFERENCE=master\nIncremental Build:\t\tdisabled\nRemove Old Build:\t\tdisabled\nBuilder Pull Policy:\t\tif-not-present\nPrevious Image Pull Policy:\talways\nQuiet:\t\t\t\tdisabled\nLayered Build:\t\t\tdisabled\nWorkdir:\t\t\t/tmp/s2i-build760427484\nDocker NetworkMode:\t\tcontainer:fb17bc2d26d6ae0bbe247e6212253032bfd5c226bfef4f24a558ddbaaaa61f18\nDocker Endpoint:\t\tunix:///var/run/docker.sock\n"
I0805 16:09:38.872700       1 docker.go:355] Using locally available image "openshift/nodejs-010-centos7@sha256:851ec2ace8a73a9cc3868edcbff1ac50d5f382d5079b7bfa4e7ff2bb531c171f"
I0805 16:09:38.893547       1 docker.go:355] Using locally available image "openshift/nodejs-010-centos7@sha256:851ec2ace8a73a9cc3868edcbff1ac50d5f382d5079b7bfa4e7ff2bb531c171f"
I0805 16:09:38.893611       1 docker.go:475] Image contains io.openshift.s2i.scripts-url set to 'image:///usr/libexec/s2i'
I0805 16:09:38.893853       1 sti.go:140] Preparing to build zupme/nodejs-teste-1:dcfd037c
I0805 16:09:38.901151       1 source.go:197] Downloading "https://github.com/openshift/nodejs-ex.git" ...
I0805 16:09:54.901598       1 cleanup.go:23] Removing temporary directory /tmp/s2i-build760427484
I0805 16:09:54.901765       1 fs.go:156] Removing directory '/tmp/s2i-build760427484'
F0805 16:09:54.902424       1 builder.go:204] Error: build error: timeout while waiting for remote repository "https://github.com/openshift/nodejs-ex.git"`

**Comment by bkmagnetron** [comment]
@EddFigueiredo  adding `--dns x.x.x.x  --dns x.x.x.x` to docker file before installation will be removed after installation, try adding it again and restart the nodes. And also there is no need to add `--dns x.x.x.x  --dns x.x.x.x` when `NM_CONTROLLED=yes`. So, I'm not sure what exactly the issue is. Follow this [link](https://docs.openshift.org/latest/install_config/install/prerequisites.html#prereq-dns).

**Comment by sdodson** [comment]
@EddFigueiredo That looks like you have a wildcard dns entry for a zone that's also part of your search path. That's a bad idea. The resolver will use the search path for any hostname that has less than 5 dots, (see man 5 resolv.conf entry for ndots). That means that it will attempt to resolve www.google.com.$project.cluster.local www.google.com.cluster.local www.google.com.local and then www.google.com.openstack.local which you've got a wildcard for openstack.local that resolves to 92.242.140.6. Is that true? Can you remove that wildcard and see if the situation improves?

**Comment by EddFigueiredo** [comment]
@sdodson Actually, my DNS wildcard is configured to another domain 

> *.openshift-testbed.local.zup.com.br

After deployed, the resolv.conf on the master node changed to 

> search openstacklocal novalocal

novalocal is the default localdomain for the environment. 

I must be doing something very wrong, cause I'm trying to make this work for about 2 weeks now, and I have a deadline on the company and I'm beginning to loose my sanity.
